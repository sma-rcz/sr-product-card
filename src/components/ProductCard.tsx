import React, { createContext } from 'react'
import styles from '../styles/styles.module.css'
import { useProduct } from '../hooks/useProduct'
import { InitialValues, onChangeArgs, Product, ProductCardHadnlers, ProductContextProps } from '../interfaces/interfaces';



// Contexto para compartir el estado del producto a sus hijos
export const ProductContext = createContext({} as ProductContextProps); // se inicializa con un objeto vacio y se le asigna el tipo de dato que va a tener
const { Provider } = ProductContext  // destructuring para obtener el provider





export interface Props {
  product: Product;
  //children?: ReactElement | ReactElement[];
  children: (args:ProductCardHadnlers) => JSX.Element;
  className?: string; // se define el tipo de la propiedad para poder agrer stylos
  style?:React.CSSProperties; // se necesita para poder agregar estilos al componente React.CSSProperties
  onChange?: (args:onChangeArgs) => void; // se define el tipo de la propiedad para poder agrer stylos
  value?:number;
  initialValues?:InitialValues
}



// Componente que renderiza la tarjeta del producto
export const ProductCard = ({ children, product,className ,style, onChange ,value,initialValues}: Props) => {
  const { counter, increseBy, maxCount
    , isMaxCountReached, reset} = useProduct(
    {
      onChange, // se le pasa la funcion onChange al hook
      product, // se le pasa el producto al hook
      value,
      initialValues,
      
   

  });


  return (
    <>{/* se envuelve todo en el provider para que los hijos puedan acceder al estado del producto */}
      <Provider value={{

        counter,
        increseBy,
        product,
        maxCount,
      }}>
        <div 
        style={style}
        className={`${styles.productCard} ${className}`}>

          {/* <span className={styles.productDescription}>{product.title}</span>*/}

          {children({
            count:counter,
            isMaxCountReached,
            maxCount:maxCount,
            product,
            increseBy,
            reset
          })}



        </div>
      </Provider>

    </>

  )
}

