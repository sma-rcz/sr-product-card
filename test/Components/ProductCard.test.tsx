import renderer from 'react-test-renderer';
import ProductCard from '../../src/components';
import { product } from '../data/products';
import React from 'react';

const  {act} = renderer;
describe('Pruebas en el ProductCard', () => { 
    test('debe de mostrar el componenete correctamente ', () => {
        const wrapper =  renderer.create (
            <ProductCard product={product}> 
                {
                    ()=> (
                        <h1></h1>
                    )
                }
            </ProductCard>
        )
        expect(wrapper.toJSON()).toMatchSnapshot();
      
    });

    test('debe de incrementar el contador ', () => {

        const wrapper = renderer.create(
            <ProductCard product={product}>
                {
                    ({count,increseBy})=> ( 
                        <>
                        <h1>Product Card</h1>
                        <span>{count}</span>
                        <button onClick={()=> increseBy(2)}></button>
                        </>
                    )
                }
            </ProductCard>
        

        )
        let tree = wrapper.toJSON();
        //console.log(tree);
        expect(tree).toMatchSnapshot();
    

      //  (tree as any).children[2].props.onClick();
      act(()=> (tree as any).children[2].props.onClick());

        tree = wrapper.toJSON();
        //console.log((tree as any).children[1].children[0])
        expect(((tree as any).children[1].children[0])).toBe("2");


    });

    
 })