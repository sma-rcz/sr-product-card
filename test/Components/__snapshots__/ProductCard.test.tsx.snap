// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el ProductCard debe de incrementar el contador  1`] = `
<div
  className="productCard undefined"
>
  <h1>
    Product Card
  </h1>
  <span>
    0
  </span>
  <button
    onClick={[Function]}
  />
</div>
`;

exports[`Pruebas en el ProductCard debe de mostrar el componenete correctamente  1`] = `
<div
  className="productCard undefined"
>
  <h1 />
</div>
`;
